openwisp_users/backends.py

```diff
@@ -0,0 +1,16 @@
+from django.contrib.auth import get_user_model
+from django.contrib.auth.backends import ModelBackend
+from django.db.models import Q
+
+User = get_user_model()
+
+
+class UserAuthenticationBackend(ModelBackend):
+    def authenticate(self, request, username=None, password=None, **kwargs):
+        users = User.objects.filter(
+            Q(email=username) | Q(phone_number=username) | Q(username=username)
+        )
+        for user in users:
```

Comment:
this looks like a potential security issue, I think we should just get the first result of the queryset

Reply:
@nemesisdesign Please which security issue are talking about?
I don't think getting the first result of the queryset will be a good idea because, if we consider the case below:

```
user1 = self._create_user(
            username='tester',
            email='tester@gmail.com',
            phone_number='+237675579231',
            password='tester',
        )
        user2 = self._create_user(
            username='tester@gmail.com',
            email='tester1@gmail.com',
            phone_number='+237675579232',
            password='tester1',
        )
```

with the solution you proposed, 
`user2` will never be able to login if he provides `username and password` as credentials. this is because `user1` will be the first in the queryset since we are checking for email first. 

I believe user `passwords` are secret, that is the reason why in my solution, I marked a user as authenticated only if the password supplied matches with the user's password and the user is active. 

Another solution I thought about is adding a clean to `User model` which prevent users from signing up with a `username` which is either the `phone_number` or `email` of already existing user.

If these solutions are not good enough, Please explain to me the security issues you saw and why you think getting the first result is the best thing to do

Reply:
`user2` is trying to pose as user1, such a user is not supposed to exist so it's not a problem at all if it can't log in.
This would also simplify the code which would not need a loop.